### PR TITLE
Approvers: Add AutoApproveUnownedSubfolders option

### PR DESCRIPTION
**Pull request by alvaroaleman**
This change adds an allow_folder_creation option to the approvers. When
opted-in, it will not require approval for subfolders in a folder with
this set. It is implemented by purging the files that fall into this
category from the list of approval-needing files in the approve plugin.

Note for reviewers: In two occasions I have changed a set of table-driven tests to use subtests, please disable whitespace on the diff, that drastically reduces the diff size.

/assign @stevekuznetsov @cjwagner

### CODE REVIEW

File: prow/repoowners/repoowners.go
Comment (fejta) [inline prow/repoowners/repoowners.go:52]: Can we choose a different name here? `AutoApprove`?

I don't understand why "allow folder creation" would mean "auto approve files under this folder" or "do not require files under here to need approval"

File: prow/plugins/approve/approvers/owners.go
Comment (fejta) [inline prow/plugins/approve/approvers/owners.go:188]: parent

**Review by fejta — COMMENTED** [review]
Nice feature. Can I bike shed this name slightly?

File: prow/repoowners/repoowners.go
Comment (fejta) [reply prow/repoowners/repoowners.go:52]: AutoAllow, AllowFolder, whatever. I'm primarily getting caught on the "Creation" word

Comment (alvaroaleman) [reply prow/repoowners/repoowners.go:52]: Generally I am definitely open for suggestions on how to improve the name. The Creation stems from the fact that this option is not supposed to get inherited, i.E. if you have folder/OWNERS with this enabled and folder/subfolder/OWNERS, then changes to folder/subfolder should _not_ get autoapproved. 

Maybe `AutoApproveSubfoldersIfNoOwners`?

Comment (fejta) [reply prow/repoowners/repoowners.go:52]: So if I set AllowFolderCreation for folder/OWNERS then folder/foo/whatever.go gets autoapproved because there is no folder/foo/OWNERS whereas folder/bar/something.go needs approval from someone (/OWNERS, folder/OWNERS, folder/bar/OWNERS) because folder/bar/OWNERS exists?

And furthermore that folder/bar/more/this.go will need the same approval (unless folder/bar/OWNERS or folder/bar/more/OWNERS.go sets this flag?)

Comment (alvaroaleman) [reply prow/repoowners/repoowners.go:52]: Yes, exactly. The underlying usecase is to allow ppl to PR new config in without requiring approval but not to allow them to mess with other ppls config. In most cases this will be completed by additional verification that checks that the new config doesn't override existing config. It would for example help with the delegation story in https://docs.google.com/document/d/1S8DFXrvFw27xYyHMDZmtUkVvd0cbDzFI8T5H35FkCyU/edit# (currently listed as "possibly openshift-specific" but it appears that ppl think it's a good idea to do it upstream)

Comment (fejta) [reply prow/repoowners/repoowners.go:52]: Gotcha! This feature sounds great.

Comment (fejta) [reply prow/repoowners/repoowners.go:52]: I do not at all understand the link between "Creation" and "Not inherited by children." Can we drop that word?

Personally I think just documenting that a child OWNERS file (and the files owned by that child file) does not inherit this feature is sufficient (without adding Creating to the word).

Your doc comment calls this "auto-approved", how about using that word?

AKA `AutoApprove` `AutoApproveSubfolders` or `AutoApproveChildren` whatever.

**Review by fejta — APPROVED** [review]
/hold

File: prow/repoowners/repoowners.go
Comment (alvaroaleman) [reply prow/repoowners/repoowners.go]: Renamed to `AutoApproveSubfolders`

**Comment by alvaroaleman** [comment]
` unable to save remote image gcr.io/k8s-testimages/launcher.gcr.io/google/bazel@sha256:e006f1c3658dd11d5176d5f7a862df4d5b9c06cfe014d5b8a86bb64b20a6f8be: unable to write image layers:`

/retest

**Comment by alvaroaleman** [comment]
/retest

Same error again

**Comment by alvaroaleman** [comment]
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from @io_k8s_repo_infra//hack:verify-boilerplate
-----------------------------------------------------------------------------
2 files have incorrect boilerplate headers:
prow/test/integration/test/crier_test.go
prow/test/integration/test/overall_test.go
```

The heck, this PR doesn't touch those files at all
/retest

**Comment by alvaroaleman** [comment]
@chaodaiG I am getting the above-mentioned error even on retest, any idea what is going on? This change doesn't touch the files the check complains about at all

**Comment by chaodaiG** [comment]
> @chaodaiG I am getting the above-mentioned error even on retest, any idea what is going on? This change doesn't touch the files the check complains about at all

This is the same error Cole encountered https://github.com/kubernetes/test-infra/pull/20656#issuecomment-770113854. I have looked on Cole's PR but didn't get any clue

**Comment by chaodaiG** [comment]
Something mystery going on here. The PR that I reverted https://github.com/kubernetes/test-infra/pull/20759/files#diff-725d3ad3307a7361f8b1f5e1d815433c31333c8511cf142a7f46ce3045c24184R8 apparently touched the boilerplate of these 2 files, and passed presubmit test, so either the one before or the one after is correct, or both are correct?

**Comment by chaodaiG** [comment]
very confusing, how did https://github.com/kubernetes/test-infra/pull/20759 pass while this one doesn't

**Comment by alvaroaleman** [comment]
/retest
The boilerplate issue was fixed in https://github.com/kubernetes/test-infra/pull/20763

**Comment by alvaroaleman** [comment]
image pull failure again
/retest

**Review by fejta — APPROVED** [review]

File: prow/repoowners/repoowners.go
Comment (alvaroaleman) [reply prow/repoowners/repoowners.go]: I've renamed it yet again to `AutoApproveUnownedSubfolders` to clarify that it will not apply when a subfolder has an owners file, WDYT?

**Pull request merged**
